### PR TITLE
Deployer

### PR DESCRIPTION
**Pull request by CyclopsRex**

### CODE REVIEW

**Review by CyclopsRex — COMMENTED** [review]
LGTM, but I think we are missing:

 - A constructor fo the Deployer that takes MaterVAMM, MasterMarginEngine, and `address newFactoryOwner` as parameters.
  - (optional) would also be nice to add the quorum and timelock periods, and the NFT address, as params in a Deployer constructor. Making all of these settable in the constructor would make it easier / less error prone to test on kovan with different settings.
 - logic to reassign ownership
 - (optional) events
 - A script to deploy the MasterVAMM, MasterMarginEngine, and Deployer.
 
 Even with the changes, this contract has some limitations that are probably fine but are nonetheless worth noting:
  - token holders can only vote once
  - token holders have to vote (and pay gas) once per NFT
  - by storing `hasTokenIdVoted` as an NFT, gas costs will be higher than if this were stored as an array (array data would be packed but I think mapping data will not be; both options are less efficient but simpler than a bitmap.) But simplicity should be our priority given our timescales.
  - tokens can be transferred before/during the voting period and this affects who can vote. I think this is fine but it goes against Simon's suggestion that we use a snapshot.

File: contracts/deployer/CommunityDeployer.sol [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+contract CommunityDeployer {
+    /// @notice The number of votes in support of a proposal required in order for a quorum to be reached and for a vote to succeed
+    uint256 public constant QUORUM_VOTES = 1; // TODO: make sure to change before deployment!!!
+
+    /// @notice Voting Period In Seconds, i.e. after 2 days elapse since the deployement of this contract, nft holders won't be able to vote
+    uint256 public constant VOTING_PERIOD_IN_SECONDS = 172800; // 2 days
+
+    /// @notice Timelock Period In Seconds, once the deployment is queued, 2 days need to pass in order to make deployment of the Voltz Factory possible
+    uint256 public constant TIMELOCK_PERIOD_IN_SECONDS = 172800; // 2 days
```
[the remark this answers is not in the document]
Comment (arturbeg) [reply contracts/deployer/CommunityDeployer.sol:40]: oh, nice

File: contracts/deployer/CommunityDeployer.sol [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+// add events
+// verify with etherscan
+// we are unable to deploy both the master vamm and the master margin engine in this contract since in that scenario it would
+// exceed the maximum contract size limit, instead we deploy the master margin engine and master vamm separately and link their addresses
+// to the community deployer as constants
+// todo: need to agree on the quorum
+
+contract CommunityDeployer {
+    /// @notice The number of votes in support of a proposal required in order for a quorum to be reached and for a vote to succeed
+    uint256 public constant QUORUM_VOTES = 1; // TODO: make sure to change before deployment!!!
```
[the remark this answers is not in the document]
Comment (arturbeg) [reply contracts/deployer/CommunityDeployer.sol:34]: will make these arguments for the constructor to be safe

File: contracts/deployer/CommunityDeployer.sol [hunk 3]
```diff
@@ -7,35 +7,39 @@ import "../Factory.sol";
...
 
-contract CommunityDeployer {
-    /// @notice The number of votes in support of a proposal required in order for a quorum to be reached and for a vote to succeed
-    uint256 public constant QUORUM_VOTES = 1; // TODO: make sure to change before deployment!!!
+// todo: deploy on kovan
+// todo: adjust the webflow ui
+// todo: investigate https://github.com/Uniswap/merkle-distributor/tree/master/scripts
+// todo: do we want to check if you still hold the nft after the snapshot?
+// todo: where do I get the index from?
+// todo: verify with etherscan
```
Comment (CyclopsRex) [inline contracts/deployer/CommunityDeployer.sol:23]: Are these todos done?

File: contracts/deployer/CommunityDeployer.sol [hunk 4]
```diff
@@ -58,8 +62,44 @@ contract CommunityDeployer {
...
 
-    constructor() {
+    // Merkle Tree
+    bytes32 public merkleRoot;
+
+    // This is a packed array of booleans.
+    mapping(uint256 => uint256) private votedBitMap;
+
+    // This event is triggered whenever a call to cast a vote succeeds
+    event Voted(uint256 index, address account, uint256 numberOfVotes); // todo: write a unit test
```
Comment (CyclopsRex) [inline contracts/deployer/CommunityDeployer.sol:72]: todo

File: contracts/deployer/CommunityDeployer.sol [hunk 5]
```diff
@@ -70,12 +110,13 @@ contract CommunityDeployer {
             "timelock is ongoing"
         );
         voltzFactory = new Factory(masterMarginEngine, masterVAMM);
+        Ownable(address(voltzFactory)).transferOwnership(ownerAddress); // todo: write a unit test
```
Comment (CyclopsRex) [inline contracts/deployer/CommunityDeployer.sol:113]: todo

File: contracts/deployer/CommunityDeployer.sol [hunk 6]
```diff
@@ -85,27 +126,44 @@ contract CommunityDeployer {
...
+            yesVoteCount += _numberOfVotes;
         } else {
-            noVoteCount++;
+            noVoteCount += _numberOfVotes;
         }
 
-        // update the hasTokenIdVoted mapping to ensure the token id cannot be reused for a duplicate vote
-        hasTokenIdVoted[_tokenId] = true;
+        // emit an event
+        emit Voted(_index, msg.sender, _numberOfVotes);
```
Comment (CyclopsRex) [inline contracts/deployer/CommunityDeployer.sol:167]: Probably useful to know which way they voted :)

**Review by CyclopsRex — COMMENTED** [review]
@arturbeg contracts lgtm. Some minor comments.

Lots of new script code here for the merkle tree generation and its tests, and I haven't taken the time to go through that in great detail. I'm guessing that a lot of this is code that has been adapted from a trusted source rather than code you've written from scratch?

If so and you'd like me review your adaptations, please point me at a diff vs. the original.

**Pull request merged**